src/components/dialog/dialog.ts

```diff
@@ -0,0 +1,260 @@
+    name: string,
+    oldVal: string
+  ) {
+    if (this.hasAttribute(attribute)) {
+      this.removeAttribute(attribute);
+    }
+
+    this.requestUpdate(name, oldVal);
+  }
+
+  private handleCancel(event: Event) {
+    event.preventDefault();
+
+    if (this.closeOnEscape) {
+      this.hide();
+    }
+  }
+
+  private handleClick(ev: MouseEvent) {
+    const elements = ev
+      .composedPath()
+      .filter((e) => e instanceof HTMLElement)
+      .map((e) => e as HTMLElement);
+    const firstElement = elements[0];
+    const firstElementRect = firstElement.getBoundingClientRect();
+    const dialogElement = elements.filter(
+      (e) => e.tagName.toLowerCase() === 'dialog'
+    )[0];
+    const dialogElementRect = dialogElement.getBoundingClientRect();
+
+    let clientX = ev.clientX;
+    let clientY = ev.clientY;
+    if (firstElement !== dialogElement) {
+      clientX = Math.max(ev.clientX, firstElementRect.left);
+      clientY = Math.max(ev.clientY, firstElementRect.top);
+    }
+
+    const clickedInside =
+      dialogElementRect.top <= clientY &&
+      clientY <= dialogElementRect.bottom &&
+      dialogElementRect.left <= clientX &&
+      clientX <= dialogElementRect.right;
+
+    if (!clickedInside && this.closeOnOutsideClick) {
+      this.hide();
+    }
+  }
+
+  private handleOpening() {
+    const args = { cancelable: true };
+    return this.emitEvent('igcOpening', args);
+  }
+
+  private handleClosing(): boolean {
+    const args = { cancelable: true };
+    return this.emitEvent('igcClosing', args);
+  }
+
+  private handleSlotChange(event: any) {
+    const elements = event.target.assignedNodes({ flatten: true });
+    elements.forEach((element: any) => {
+      if (element.querySelector) {
+        const form =
+          element.querySelector('form') || element.querySelector('igc-form');
+        if (form && form.getAttribute('method') === 'dialog') {
+          const submitEvent =
+            form.tagName.toLowerCase() === 'form' ? 'submit' : 'igcSubmit';
+          form.addEventListener(submitEvent, (ev: any) => {
+            const submitter = submitEvent === 'submit' ? ev.submitter : null;
+            this.returnValue = submitter ? submitter.value : '';
+            this.hide();
+          });
+
+          return;
+        }
+      }
+    });
+  }
+
+  protected override render() {
+    return html`
+      <div part="backdrop" aria-hidden="true" ?hidden=${!this.open}></div>
+      <dialog
+        part="base"
+        @click=${this.handleClick}
+        @cancel=${this.handleCancel}
+        aria-label=${ifDefined(this.ariaLabel)}
+      >
+        <header part="title" id=${this.titleId}>
+          <slot name="title"><span>${this.title}</span></slot>
+        </header>
+        <section part="content">
+          <slot @slotchange=${this.handleSlotChange}></slot>
+        </section>
+        <footer part="footer">
+          <slot name="footer"></slot>
+        </footer>
+      </dialog>
+    `;
+  }
+}
```

Comment:
```suggestion
protected override render() {
    const label = this.ariaLabel ? this.ariaLabel : undefined;
    const labelBy = label ? undefined : this.titleId;

    return html`
      <div part="backdrop" aria-hidden="true" ?hidden=${!this.open}></div>
      <dialog
        part="base"
        @click=${this.handleClick}
        @cancel=${this.handleCancel}
        aria-label=${ifDefined(label)}
        aria-labelledby=${ifDefined(labelBy)}
      >
        <header part="title" id=${this.titleId}>
          <slot name="title"><span>${this.title}</span></slot>
        </header>
        <section part="content">
          <slot @slotchange=${this.handleSlotChange}></slot>
        </section>
        <footer part="footer">
          <slot name="footer"></slot>
        </footer>
      </dialog>
    `;
  }
```
I've spent some time testing the component with NVDA and I came to the following conclusions:

- we don't need the role attribute on the component since the interactable part is the underlying dialog element which implicitly has the correct role.
- in addition, having role=dialog on the component without the rest of the aria attributes is an error.
- if `aria-label` is provided for the igc-dialog, pass it down to the native dialog and remove the labelledby attribute.
- otherwise use the labelledby attribute, since the id reference is in the same DOM tree (`titleID`)